### PR TITLE
API to validate the webui key

### PR DESCRIPTION
**Pull request by vinay033**
### :nut_and_bolt: Description: What code changed, and why?

we need to add one API to validate the webui key for the server.
### :chains: Related Resources
https://github.com/chef/automate/issues/6273
### :+1: Definition of Done
I have added API to validate the webui key.
### :athletic_shoe: How to Build and Test the Change
Steps to build:
```
rebuild components/infra-proxy-service
rebuild components/automate-gateway
```
```
webui key is available for the server - 

curl --location --request POST 'https://a2-dev.test/api/v0/infra/servers/validate' \
--header 'api-token: your admin token' \
--header 'Content-Type: text/plain' \
--data-raw '{
    "id":"chef-server",
    "fqdn":"a2-dev.test",
    "webui_key": "",
    "is_webui_key":true
}'

webui key is not available for the server going to add new one

curl --location --request POST 'https://a2-dev.test/api/v0/infra/servers/validate' \
--header 'api-token: your admin token' \
--header 'Content-Type: text/plain' \
--data-raw '{
    "id":"",
    "fqdn":"a2-dev.test",
    "webui_key": "---webui key----",
    "is_webui_key":false
}'

```
Note: 
in case of is_webui_key - true means webui key is already available for server, if is_webui_key is false means you are going to add new webui key

valid response
```
{
    "valid": true,
    "error": ""
}
```

Invalid response
```
{
    "valid": false,
    "error": "The user or client who made the request could not be authenticated. Verify the user/client name, and that the correct key was used to sign the request."
}
```
### :white_check_mark: Checklist

**All PRs** must tick these:

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [x] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [x] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [x] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [x] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [x] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [x] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### CODE REVIEW

File: api/external/infra_proxy/request/servers.proto [hunk 1]
```diff
@@ -41,3 +41,14 @@ message GetServerStatus {
     // Server FQDN
     string fqdn = 1;
 }
+
+message ValidateWebuiKey {
+    // Chef Infra Server ID.
+    string id = 1;
+    // Chef Server FQDN
+    string fqdn = 2;
+    // Chef Server Webui Key
```
Comment (kalroy) [inline api/external/infra_proxy/request/servers.proto:50]:
```suggestion
    // Optional Chef Server Webui Key
```

File: api/external/infra_proxy/request/servers.proto [hunk 2]
```diff
@@ -41,3 +41,14 @@ message GetServerStatus {
...
+    string webui_key = 3;
+    // Chef Server Webui key is availbale
+    bool is_webui_key = 4;
```
Comment (kalroy) [inline api/external/infra_proxy/request/servers.proto:53]: Do not understand the necessity of passing this?

File: api/interservice/infra_proxy/request/servers.proto [hunk 3]
```diff
@@ -43,3 +43,14 @@ message GetServerStatus {
     // Chef Infra Server IP Address.
     string ip_address = 2;
 }
+
+message ValidateWebuiKey {
+    // Chef Infra Server ID.
+    string id = 1;
+    // Chef Server FQDN
+    string fqdn = 2;
+    // Chef Server Webui Key
```
Comment (kalroy) [inline api/interservice/infra_proxy/request/servers.proto:52]:
```suggestion
    // Optional Chef Server Webui Key
```

File: api/interservice/infra_proxy/request/servers.proto [hunk 4]
```diff
@@ -43,3 +43,14 @@ message GetServerStatus {
...
+    string webui_key = 3;
+    // Chef Server Webui key is availbale
+    bool is_webui_key = 4;
```
Comment (kalroy) [inline api/interservice/infra_proxy/request/servers.proto:55]: Do not understand the need of passing it as request

File: components/automate-gateway/handler/infra_proxy/servers.go [hunk 5]
```diff
@@ -37,6 +38,26 @@ func (a *InfraProxyServer) GetServer(ctx context.Context, r *gwreq.GetServer) (*
 	}, nil
 }
 
+// ValidateWebuiKey validates the webui key
+func (a *InfraProxyServer) ValidateWebuiKey(ctx context.Context, r *gwreq.ValidateWebuiKey) (*gwres.ValidateWebuiKey, error) {
+	req := &infra_req.ValidateWebuiKey{
+		Id:         r.Id,
+		Fqdn:       r.Fqdn,
+		WebuiKey:   r.WebuiKey,
+		IsWebuiKey: r.IsWebuiKey,
```
Comment (kalroy) [inline components/automate-gateway/handler/infra_proxy/servers.go:47]: We should pass it as true and no need to get it from request:
```suggestion
		IsWebuiKey: true,
```

File: components/infra-proxy-service/server/proxy.go [hunk 6]
```diff
@@ -160,6 +162,19 @@ func GetOrgAdminKeyFrom(secret *secrets.Secret) string {
 	return adminKey
 }
 
+// GetWebuiKeyFrom returns Webuikey
```
Comment (kalroy) [inline components/infra-proxy-service/server/proxy.go:165]: The code is already written `GetOrgAdminKeyFrom`
We need to do this:
- Refactor the code to a common function
- Use the common function in both these functions

File: components/infra-proxy-service/server/servers.go [hunk 7]
```diff
@@ -63,6 +64,45 @@ func (s *Server) CreateServer(ctx context.Context, req *request.CreateServer) (*
 	}, nil
 }
 
+// ValidateWebuiKey validate the webui key
+func (s *Server) ValidateWebuiKey(ctx context.Context, req *request.ValidateWebuiKey) (*response.ValidateWebuiKey, error) {
+
+	if req.IsWebuiKey && req.Id != "" {
```
Comment (kalroy) [inline components/infra-proxy-service/server/servers.go:70]: We do not need `req.IsWebuiKey`. We can check `req.WebUIKey` for empty string

File: components/infra-proxy-service/server/servers.go [hunk 8]
```diff
@@ -63,6 +64,45 @@ func (s *Server) CreateServer(ctx context.Context, req *request.CreateServer) (*
...
+			return nil, service.ParseStorageError(err, *req, "server")
+		}
+
+		// Get web ui key from secrets service
+		secret, err := s.service.Secrets.Read(ctx, &secrets.Id{Id: server.CredentialID})
+		if err != nil {
+			return nil, err
+		}
+
+		req.WebuiKey = GetWebuiKeyFrom(secret)
```
Comment (kalroy) [inline components/infra-proxy-service/server/servers.go:82]: Never modify request object.
Declare a local variable in the function to store the passed the WebUIKey and change it

File: components/infra-proxy-service/server/servers.go [hunk 9]
```diff
@@ -63,6 +64,45 @@ func (s *Server) CreateServer(ctx context.Context, req *request.CreateServer) (*
...
+func (s *Server) ValidateWebuiKey(ctx context.Context, req *request.ValidateWebuiKey) (*response.ValidateWebuiKey, error) {
+
+	if req.IsWebuiKey && req.Id != "" {
+		server, err := s.service.Storage.GetServer(ctx, req.Id)
+		if err != nil {
+			return nil, service.ParseStorageError(err, *req, "server")
+		}
+
+		// Get web ui key from secrets service
+		secret, err := s.service.Secrets.Read(ctx, &secrets.Id{Id: server.CredentialID})
```
Comment (kalroy) [inline components/infra-proxy-service/server/servers.go:77]: We need to also add another condition to check if the server table has credential_id or not.
If not, we should return false and error

File: components/automate-gateway/handler/infra_proxy/servers.go [hunk 10]
```diff
@@ -37,6 +38,26 @@ func (a *InfraProxyServer) GetServer(ctx context.Context, r *gwreq.GetServer) (*
...
+	}
+
+	res, err := a.client.ValidateWebuiKey(ctx, req)
+	if err != nil {
+		log.Warnf("Validate webui key error:: %s", err.Error())
+		return nil, err
```
Comment (kalroy) [inline components/automate-gateway/handler/infra_proxy/servers.go:53]: In case of error we should return the response with false validation

**Review by kalroy — CHANGES_REQUESTED** [review]

**Comment by sonali523** [comment]
While testing, it's observed that sometimes if we change any character in web ui key then also the API is giving 
{
    "valid": true,
    "error": ""
}

**Review by kalroy — APPROVED** [review]

**Review by sonali523 — APPROVED** [review]

**Pull request merged**